src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/DeploymentPosterImpl.java

```diff
@@ -0,0 +1,82 @@
+import com.atlassian.bitbucket.jenkins.internal.credentials.JenkinsToBitbucketCredentials;
+import com.atlassian.bitbucket.jenkins.internal.model.deployment.BitbucketCDCapabilities;
+import com.atlassian.bitbucket.jenkins.internal.model.deployment.BitbucketDeployment;
+import com.cloudbees.plugins.credentials.Credentials;
+import hudson.model.Run;
+import hudson.model.TaskListener;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.util.Optional;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+@Singleton
+public class DeploymentPosterImpl implements DeploymentPoster {
+
+    protected static final Logger LOGGER = Logger.getLogger(DeploymentPosterImpl.class.getName());
+
+    @Inject
+    private BitbucketClientFactoryProvider bitbucketClientFactoryProvider;
+    @Inject
+    private JenkinsToBitbucketCredentials jenkinsToBitbucketCredentials;
+    @Inject
+    private BitbucketPluginConfiguration pluginConfiguration;
+
+    @Override
+    public void postDeployment(String bbsServerId, String projectKey, String repositorySlug, String revisionSha,
+                               BitbucketDeployment deployment, Run<?, ?> run,
+                               TaskListener taskListener) {
+        Optional<BitbucketServerConfiguration> maybeServer = pluginConfiguration.getServerById(bbsServerId);
+        if (!maybeServer.isPresent()) {
+            taskListener.error(String.format("Could not send deployment notification to Bitbucket Server: Unknown serverId %s", bbsServerId));
+            return;
+        }
+
+        BitbucketServerConfiguration server = maybeServer.get();
+        BitbucketClientFactory clientFactory = getClientFactory(run, server);
+        BitbucketCDCapabilities cdCapabilities = clientFactory.getCapabilityClient().getCDCapabilities();
+        if (!cdCapabilities.supportsDeployments()) {
+            // Bitbucket doesn't have deployments
+            taskListener.error("Could not send deployment notification to Bitbucket Server: The Bitbucket version does not support deployments");
+            return;
+        }
+
+        taskListener.getLogger().printf("Sending notification of %s to %s on commit %s%n",
+                deployment.getState().name(), server.getServerName(), revisionSha);
+        try {
+            clientFactory.getProjectClient(projectKey)
+                    .getRepositoryClient(repositorySlug)
+                    .getDeploymentClient(revisionSha)
+                    .post(deployment);
+            taskListener.getLogger().printf("Successfully sent notification of %s deployment to %s on commit %s%n",
+                    deployment.getState().name(), server.getServerName(), revisionSha);
+        } catch (AuthorizationException e) {
+            taskListener.error(String.format("Failed to send notification of deployment to %s due to an authorization error: %s",
+                    server.getServerName(), e.getMessage()));
```

Comment:
`e.getMessage` is always  "Provided credentials cannot access the resource" So this will read
"Failed to send notification of deployment to {serverName} due to an authorization error: Provided credentials cannot access the resource" it feels long and clumsy for saying "Couldn't access the resource, check your password". Maybe we can make it better? what should I do if this happens in my build? what credentials are broken?

Reply:
We don't have a lot of visibility into which credentials are broken, as we get them from:
```java
GlobalCredentialsProvider globalCredentialsProvider = server.getGlobalCredentialsProvider(run.getParent());
Credentials globalAdminCredentials = globalCredentialsProvider.getGlobalAdminCredentials().orElse(null);
BitbucketCredentials credentials = jenkinsToBitbucketCredentials.toBitbucketCredentials(globalAdminCredentials);
```
Which means we may be making an anonymous call if there are no global credentials for the run. Additionally, `BitbucketCredentials` don't have any kind of id/name or anything else that we can use to identify it and neither do `Credentials`.

Reply:
That says exactly which credentials it is, it is the "credentials associated with the Bitbucket server instance. Please contact your admin to update the credentials"
Because you can only have one token per BbS instance and we always use that one to the exclusion of all others you know exactly which one it is and can provide guidance on how to fix it.